src/ynEIGEN/AssetRegistry.sol

```diff
@@ -0,0 +1,342 @@
+
+    error UnsupportedAsset(IERC20 asset);
+    error Paused();
+    error AssetNotActiveOrNonexistent(address inactiveAsset);
+    error AssetBalanceNonZeroInPool(uint256 balanceInPool);
+    error AssetBalanceNonZeroInStrategyManager(uint256 balanceInStrategyManager);
+    error AssetNotFound(address absentAsset);
+    error ZeroAmount();
+    error ZeroAddress();
+    error LengthMismatch(uint256 length1, uint256 length2);
+    error AssetAlreadyActive(address asset);
+    error AssetAlreadyInactive(address asset);
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  ROLES  -------------------------------------------
+    //--------------------------------------------------------------------------------------
+
+    bytes32 public constant ASSET_MANAGER_ROLE = keccak256("ASSET_MANAGER_ROLE");
+    bytes32 public constant PAUSER_ROLE = keccak256("PAUSER_ROLE");
+    bytes32 public constant UNPAUSER_ROLE = keccak256("UNPAUSER_ROLE");
+    
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  VARIABLES  ---------------------------------------
+    //--------------------------------------------------------------------------------------
+
+     /// @notice List of supported ERC20 asset contracts.
+    IERC20[] public assets;
+
+    mapping(IERC20 => AssetData) public _assetData;
+
+    bool public actionsPaused;
+
+    IRateProvider public rateProvider;
+    IEigenStrategyManager public eigenStrategyManager;
+    IynEigen ynEigen;
+
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  INITIALIZATION  ----------------------------------
+    //--------------------------------------------------------------------------------------
+
+    constructor() {
+       _disableInitializers();
+    }
+
+    struct Init {
+        IERC20[] assets;
+        IRateProvider rateProvider;
+        IEigenStrategyManager eigenStrategyManager;
+        IynEigen ynEigen;
+        address admin;
+        address pauser;
+        address unpauser;
+    }
+
+    function initialize(Init calldata init)
+        public
+        notZeroAddress(address(init.rateProvider))
+        notZeroAddress(address(init.admin))
+        notZeroAddress(address(init.pauser))
+        notZeroAddress(address(init.unpauser))
+        initializer {
+        __AccessControl_init();
+        __ReentrancyGuard_init();
+
+        _grantRole(DEFAULT_ADMIN_ROLE, init.admin);
+        _grantRole(PAUSER_ROLE, init.pauser);
+        _grantRole(UNPAUSER_ROLE, init.unpauser);
+
+        uint256 assetsLength = init.assets.length;
+        for (uint256 i = 0; i < assetsLength; i++) {
+            if (address(init.assets[i]) == address(0)) {
+                revert ZeroAddress();
+            }
+            assets.push(init.assets[i]);
```

Comment:
I know I know, this is called by us since its the `initialize` function.

However, there is nothing preventing duplicate assets in `init.assets`. In case of a duplicate, a duplicate will be added to the `assets` array (which is not good), and the below will simply overwrite with the same data (so thats fine). So why is it a problem: it will mess up the `deleteAsset` functioning due to there being 2 entries in the `assets` array but only one in the `_assetData` mapping.

```solidity
_assetData[init.assets[i]] = AssetData({
    active: true
});
```

So add in this line right after line 110:

```solidity
if (_assetData[init.assets[i].active) {
    revert AssetAlreadyActive(address(init.assets[i));
}
```

Reply:
actually its even more problematic since it'll mess up the `totalAssets` function and count these duplicate assets balance twice (or more depending on number of duplicates).

Reply:
likelihood: low
impact: high
severity: medium